servlet/src/main/java/io/undertow/servlet/handlers/FilterHandler.java

```diff
@@ -106,8 +106,8 @@ public void doFilter(final ServletRequest request, final ServletResponse respons
 
 
             final ServletRequestContext servletRequestContext = exchange.getAttachment(ServletRequestContext.ATTACHMENT_KEY);
-            final ServletRequest oldReq = servletRequestContext.getServletRequest();
-            final ServletResponse oldResp = servletRequestContext.getServletResponse();
+            final ServletRequest oldReq = servletRequestContext.getOriginalRequest();
+            final ServletResponse oldResp = servletRequestContext.getOriginalResponse();
```

Comment:
I really don't think this is correct. If you have multiple filters that wrap the request it will be reset to the original one after the last filter in the chain returns.

Reply:
I get your point, this is not correct.

However, the reset of request and response reference on the ServletRequestContext does not work well with some filters (like org.springframework.web.filter.ShallowEtagHeaderFilter) that wrap the response in the initial filter execution and require a response of the wrapped type when the asynchronous request completes.

I believe that the ServletRequestContext request and response reset in the finally block (https://github.com/sermojohn/undertow/blob/d37402476127c50ace20a340f64718bc8ba17d90/servlet/src/main/java/io/undertow/servlet/handlers/FilterHandler.java#L144) could be totally removed. This would result in leaking the response wrappers, but this should not affect the filters themselves because they only use the request and response arguments they receive.

What do you think?

Reply:
I think that the correct way to handle this is to store the current request/response when the AsyncContext is created.